plugins/inputs/t128_graphql/query_builder.go

```diff
@@ -1,20 +1,107 @@
 package t128_graphql
 
-import "strings"
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+
+	"github.com/Jeffail/gabs"
+)
 
 func buildQuery(entryPoint string, fields map[string]string, tags map[string]string) string {
-	var replacer = strings.NewReplacer("[", "(", "]", "\")", "/", "{", ":", ":\"")
-	query := "query {" + replacer.Replace(entryPoint) + "{"
+	query := "query "
+
+	var buf bytes.Buffer
+	w := io.Writer(&buf)
+	jsonObj := buildQueryObject(entryPoint, fields, tags)
+	traverseQueryObjPreOrder(jsonObj, w)
+
+	query += buf.String()
+	return query
+}
 
-	for _, element := range fields {
-		query += "\n" + element
+func traverseQueryObjPreOrder(jsonObj *gabs.Container, w io.Writer) {
+	jsonChildren, err := jsonObj.ChildrenMap()
+	if err != nil {
+		fmt.Println("error")
+		return
 	}
-	query = strings.TrimSpace(query)
-	for _, element := range tags {
-		query += "\n" + element
+
+	//store and sort the keys
+	keys := make([]string, len(jsonChildren))
+	i := 0
+	for k := range jsonChildren {
+		keys[i] = k
+		i++
 	}
+	sort.Strings(keys)
 
-	query += "\n" + strings.Repeat("}", strings.Count(query, "{"))
+	children := 0
+	for i, key := range keys {
+		if strings.HasPrefix(key, "$") {
+			w.Write([]byte("(" + key[1:] + ":\"" + jsonChildren[key].Data().(string) + "\"){"))
```

Comment:
Again the logic here is very dense here. Can you split it up into intermediate variable/funcs to make it more obvious what is going on?

Reply:
Meaning for all of this nested stuff, but this line more so